### DISCUSSION TITLE
How to properly release/close a podman connection?

### DISCUSSION THREAD

**Discussion by thediveo** [containers/podman#15622] [Q&A]
What is the proper way to correctly release the resources allocated to a podman connection?

I'm asking because the [Gomega go routine leak detector](https://onsi.github.io/gomega/#codegleakcode-finding-leaked-goroutines) gets tripped up. Especially by `system.Event()`, that has persistent HTTP client-related go routines left over after terminating the event stream by closing the passed-in `cancelChan`.

**Comment by vrothberg** [comment]
Thanks for reaching out, @thediveo.

I am under the impression that it might be a bug.  I'd expect resources to be cleaned up.  Would you open issues for such cases?

**Reply by thediveo** [reply]
Yes, will do. I'm not sure if I can come up with a minimal example, but in any case have an open source example with unit test that triggers the lurking go routines when using `system.Event`.

**Reply by Luap99** [reply]
What version did you test? I fixed some issue in events and logs recently https://github.com/containers/podman/pull/14984, looks like this is only in the main branch.

**Reply by thediveo** [reply]
I'm using the debian experimental podman package on Ubuntu, which is a tagged 4.2.0 IIRC.

**Reply by thediveo** [reply]
After some more tests with the unit tests (sic!) for whatever reason there are HTTPs "persistent connections" with their goroutines hanging around also after calling other REST API endpoints, so this might be due to the way the HTTP connections operate, especially in case podman tries to stick with permanent HTTP connections.

So this looks not so much as kind of "leak" but instead a need to at least in unit tests with goroutine leak checking to always get a podman's connection HTTP client and close its idle connections.

```go
conn, _ := bindings.GetClient(podconn)
conn.Client.CloseIdleConnections()

Eventually(Goroutines).WithTimeout(2 * time.Second).ShouldNot(HaveLeaked(goodgos))
```
